src/hotspot/share/nmt/mallocHeader.hpp

```diff
@@ -87,6 +88,55 @@ class outputStream;
  *   the bytes are stored individually.
  */
 
+template<typename T, typename U = T>
+class AsanPoisoningHelper {
+  U* _memory_region;
+ public:
+  AsanPoisoningHelper() = delete;
+  AsanPoisoningHelper(U* addr) : _memory_region(addr) {
```

Comment:
This is what I meant, then you don't need any casting when passing along the args.

```c++
template<typename T>
class AsanPoisoningHelper {

public:
  AsanPoisoningHelper(T* addr) : _memory_region(addr) {
    #if INCLUDE_ASAN
      ASAN_UNPOISON_MEMORY_REGION(reinterpret_cast<char*>(_memory_region), sizeof(T));
    #endif
  }
};

```

Reply:
since we are using `const ...` types for `T`, we cannot use `reinterpret_cast`. 
Casting is avoided everywhere by introducing second template parameter `U`.

Reply:
`U` is `T` with `const`, right? Just make `_memory_region` `const T*` and remove `U`. If the `reinterpret_cast` isn't needed for the macro to work, then don't use it.

If you need some more advanced modifications of the types involved, then look at the `<type_traits>` header, example: https://en.cppreference.com/w/cpp/types/remove_cv.html